sdk/core/azure-core/src/credentials/authorization_challenge_parser.cpp

```diff
@@ -0,0 +1,274 @@
+  while (TryGetNextChallenge(headerSpan, challengeKey))
+  {
+    // Enumerate each key=value parameter until we find the 'claims' key on the 'Bearer'
+    // challenge.
+    StringSpan key;
+    StringSpan value;
+    while (TryGetNextParameter(headerSpan, key, value))
+    {
+      if (challengeKey.CaseInsensitiveEquals(bearer) && key.CaseInsensitiveEquals(claims))
+      {
+        return value.ToString();
+      }
+    }
+  }
+
+  return {};
+}
+
+namespace {
+std::set<char> const Space = {' '};
+std::set<char> const SpaceOrComma = {' ', ','};
+std::set<char> const Separator = {'='};
+std::set<char> const Quote = {'\"'};
+
+bool TryGetNextChallenge(StringSpan& headerValue, StringSpan& challengeKey)
+{
+  challengeKey = StringSpan();
+
+  headerValue = headerValue.TrimStart(Space);
+  int endOfChallengeKey = headerValue.IndexOfAny(Space);
+
+  if (endOfChallengeKey < 0)
+  {
+    return false;
+  }
+
+  challengeKey = headerValue.Slice(0, endOfChallengeKey);
+
+  // Slice the challenge key from the headerValue
+  headerValue = headerValue.Slice(endOfChallengeKey + 1);
+
+  return true;
+}
+
+bool TryGetNextParameter(StringSpan& headerValue, StringSpan& paramKey, StringSpan& paramValue)
+{
+  paramKey = StringSpan();
+  paramValue = StringSpan();
+
+  // Trim any separator prefixes.
+  headerValue = headerValue.TrimStart(SpaceOrComma);
+
+  int nextSpace = headerValue.IndexOfAny(Space);
+  int nextSeparator = headerValue.IndexOfAny(Separator);
+
+  if (nextSpace < nextSeparator && nextSpace != -1)
+  {
+    // we encountered another challenge value.
+    return false;
+  }
+
+  if (nextSeparator < 0)
+  {
+    return false;
+  }
+
+  // Get the paramKey.
+  paramKey = headerValue.Slice(0, nextSeparator).Trim(Space);
+
+  // Slice to remove the 'paramKey=' from the parameters.
+  headerValue = headerValue.Slice(nextSeparator + 1);
+
+  // The start of paramValue will usually be a quoted string. Find the first quote.
+  int quoteIndex = headerValue.IndexOfAny(Quote);
+
+  // Get the paramValue, which is delimited by the trailing quote.
+  headerValue = headerValue.Slice(quoteIndex + 1);
+  if (quoteIndex >= 0)
+  {
+    // The values are quote wrapped
+    paramValue = headerValue.Slice(0, headerValue.IndexOfAny(Quote));
+  }
+  else
+  {
+    // the values are not quote wrapped (storage is one example of this)
+    //  either find the next space indicating the delimiter to the next value, or go to the end
+    //  since this is the last value.
+    int trailingDelimiterIndex = headerValue.IndexOfAny(SpaceOrComma);
+    if (trailingDelimiterIndex >= 0)
+    {
+      paramValue = headerValue.Slice(0, trailingDelimiterIndex);
+    }
+    else
+    {
+      paramValue = headerValue;
+    }
+  }
+
+  // Slice to remove the '"paramValue"' from the parameters.
+  if (!headerValue.CaseInsensitiveEquals(paramValue))
+    headerValue = headerValue.Slice(paramValue.Length() + 1);
+
+  return true;
+}
+
+StringSpan::StringSpan(std::string const* stringPtr)
```

Comment:
Why a pointer to a std::string and not a reference to a std::string? If you're trying to implement the logical equivalent of `std::string_view`, then IMHO `std::string const&` is closer to the intent (`std::string const*` is a mutable pointer to a constant std::string which is a non-obvious construct).

Reply:
Alternatively, can `StringSpan` operate on `string.begin()` and `string.end()`? That seems more in line with the existing C++ library practice.

Reply:
I thought about taking parameter by reference, and storing a pointer, but that semantics is really not obvious - when you pass something by reference and after that something holds a pointer to it. With pointer, you at least expect it more.

Reply:
For the picture, why the code semantics is largely what you see - I had an option to implement great parser for `WWW-Authenticate` from scratch, in C++ terms, or to implement a mediocre one which nevertheless covers everything we seem to need from it.

I tried to implement the complete parser, but found out it is hard to do, the grammar is ambiguous, and it would take at least few more days to get it.

Then I thought, let's start with mediocre, and maybe later improve it. But writing a mediocre paser also takes time and it may have more bugs where we don't expect it.

So I decided to take the implementation from C# SDK and make it easy to diff so that it is easy to be sure that the logic is largely the same therefore the result should be the same. We should not get more bugs than C# SDK has, and if there is one, they also have it. But they have been around longer and so far it is not known to be a source of real problems. So I took their implementation, wrote our tests, took their tests, and all is passing.

So that is what I did with this `StringSpan` class. In C++, without the `string_view` (which we can't use unfortunately because it is not C++14), you kind of need such concept anyways or you end up copying and copying the string while parsing.

Look at this diff for better illustration of what I said above: https://www.diffchecker.com/VFV9KhU9/

Reply:
I was actually thinking of saving the reference as a const reference internally rather than a pointer.

In general, I get the heeby-jeebies whenever I see a raw pointer to a c++ stl construct (I think Ahson would call it a "code smell" issue).

Reply:
I agree, but see with the limitations I had this span class is probably the most ok solution. It can be changed to store references instead of pointers, but reference data members, to me, smell more than raw pointers :)

Reply:
Plus having it as reference will make it impossible to have `operator=` for spans, which we do use.

Reply:
And this: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-constref
Not that we have to follow all the guidelines in this specific case - the class is not public, not even visible outside of the .cpp file, but in general I find it a good recommendation. Even debug-wise I find raw pointers to be bad, but reference data members way worse.